jgrapht-core/src/main/java/org/jgrapht/util/DoublyLinkedList.java

```diff
@@ -0,0 +1,485 @@
+/*
+ * (C) Copyright 2018-2018, by Timofey Chudakov and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * See the CONTRIBUTORS.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.util;
+
+import java.util.ConcurrentModificationException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * This is the implementation of the doubly linked list data structure, which gives access to the internal
+ * list nodes the data is stored in. The primary goal of this is to be able to delete an element from this
+ * list in $\mathcal{O}(1)$ via the list node it is stored in.
+ * <p>
+ * The iterators over this list have a <i>fail-fast</i> behaviour meaning that they throw a
+ * {@link ConcurrentModificationException} after they detect a structural modification of the list,
+ * that they're not responsible for.
+ *
+ * @param <E> the list element type
+ * @author Timofey Chudakov
+ */
+public class DoublyLinkedList<E> implements Iterable<E> {
+    /**
+     * The first element of the list
+     */
+    private ListNode<E> head;
+    /**
+     * The size of the list
+     */
+    private int size;
+    /**
+     * The number of modifications applied to this list. This value is used to exhibit a fail-fast
+     * behavior during the iteration over the list when its structure is changed concurrently
+     */
+    private int modCount;
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Iterator<E> iterator() {
+        return listIterator();
+    }
+
+    /**
+     * Returns a list iterator starting from the beginning of this list
+     *
+     * @return a list iterator starting from the beginning of this list
+     */
+    public ListIterator<E> listIterator() {
+        return new ListIteratorImpl(head);
+    }
+
+    /**
+     * Returns a reverse iterator over this list, which starts from the {@code element}.
+     * Note: the returned iterator will iterate over the all list, meaning it won't
+     * stop at the end of the list unless the {@code element} is the beginning of the list.
```

Comment:
I'm not sure I understand this. If an element is the first element of the list, why wouldn't the reverseIterator wrap around and walk the list back to the beginning?

Reply:
Good point, this is a bit incorrect. The reverse iterator will stop at the beginning of the list in the case the start node is the last element.

Reply:
I still find the description of `reverseIteratorFrom(.)` confusing. What about the following description:


```
Returns an iterator over this list, which traverses the list in the reverse direction. The returned iterator will iterate over the <i>entire</i> list, that is, the list is treated as if it were cyclic. The iterator starts from {@code element}, and walks the list backwards, while wrapping around at the beginning of the list. For instance, for the list [1,2,3,4], {@code reverseIteratorFrom(3)} would return 3,2,1,4.

This method throws {@link NoSuchElementException} in the case {@code element} doesn't belong to this list. This method invokes {@link #getNode(Object) getNode} to find the list node containing {@code element}.
```